libfwupdplugin/fu-input-stream.c

```diff
@@ -364,6 +364,40 @@ fu_input_stream_read_bytes(GInputStream *stream, gsize offset, gsize count, GErr
 	return g_byte_array_free_to_bytes(g_steal_pointer(&buf)); /* nocheck:blocked */
 }
 
+/**
+ * fu_input_stream_read_string:
+ * @stream: a #GInputStream
+ * @offset: offset in bytes into @stream to copy from
+ * @count: maximum number of bytes to read
+ * @error: (nullable): optional return location for an error
+ *
+ * Read a UTF-8 string from a stream in a safe way.
+ *
+ * Returns: (transfer full): string
+ *
+ * Since: 2.0.0
+ **/
+gchar *
+fu_input_stream_read_string(GInputStream *stream, gsize offset, gsize count, GError **error)
+{
+	g_autoptr(GByteArray) buf = NULL;
+
+	g_return_val_if_fail(G_IS_INPUT_STREAM(stream), NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	buf = fu_input_stream_read_byte_array(stream, offset, count, error);
+	if (buf == NULL)
+		return NULL;
+	if (!g_utf8_validate_len((const gchar *)buf->data, buf->len, NULL)) {
```

Comment:
Is there garbage collection or is "buf" going to leak ?

Reply:
I am assuming g_autoptr is autocleaning when going out of scope

Reply:
Exactly that -- g_autoptr() is a compile-time wrapper around `__attribute__((cleanup))` -- each GLib type uses `G_DEFINE_AUTOPTR_CLEANUP_FUNC` to tell the compiler what to run when the variable goes out of scope -- in this case it'd be something like `G_DEFINE_AUTOPTR_CLEANUP_FUNC(GByteArray, g_byte_array_unref)`